proposals/esm-integration/README.md

```diff
@@ -1,3 +1,308 @@
+  .instantiateStreaming(req, imports)
+  .then(
+    obj => obj.instance.exports.foo()
+  );
+```
+
+A declarative API would improve the ergonomics by making this work happen implicitly.
+
+```
+import {foo} from “./myModule.wasm”;
+foo();
+```
+
+### Enable WebAssembly modules to take part in JavaScript module graphs
+
+With the introduction of ES modules, the JS spec provides a way to express module dependency graphs.
+
+These graphs are expressed with import and export statements.
+
+```
+// main.js
+
+import {count, incrementCount} from "./dep.js"
+```
+
+```
+//dep.js
+
+let count = 10;
+
+function incrementCount() {
+    count++;
+}
+
+export {count, incrementCount};
+```
+
+WebAssembly modules currently can't take part in these graphs. This means that developers who want to use WebAssembly modules have to manually trigger instantiation, which may not have completed by the time the JS module graph returns, breaking the mental model of modules.
+
+## Proposed solution
+
+Both goals can be addressed by enabling WebAssembly modules to be loaded using the ES module system. 
+
+This system has three phases:
+
+1. Construction — Module files are fetched and parsed and the module record is constructed.
+1. Instantiation — Exports and imports are wired up to memory locations. Functions are also initialized here.
+1. Evaluation — Top-level module code is evaluated so that non-function exports can be assigned their values.
+
+The work completed in these phases is defined in two different specs:
+
+- A host-specific specification (i.e. the HTML spec) defines how to load modules.
+- The ES module spec provides methods for loaders to use to parse, instantiate, and evaluate modules. The spec defines both an Abstract Module Record and a concrete subclass of this for JS modules, called the Source Text Module Record.
+
+This proposal will create a new WebAssembly Module subclass of the Abstract Module Record. For this, the 3 phases will be specified as follows (with the possible introduction of a sub-phase in Instantiation):
+
+### Construction
+
+During construction, all files for modules in the graph are downloaded and are parsed into Module Records. 
+
+Three things happen for each module during the Construction phase.
+
+1. Finding the file — Based on the module specifier, the host figures out where to get the file from using its module resolution algorithm. Module resolution algorithms differ between hosts. There is ongoing work to fix incompatibilities between Node.js and browser module resolution with the [package-name-maps](https://github.com/domenic/package-name-maps) proposal.
+1. Loading the file — The host fetches the file using the module path generated in module resolution. For browsers, the loading algorithm is specified in the [HTML spec](https://html.spec.whatwg.org/#fetch-a-module-script-tree). The WebAssembly MIME type will need to be added to this spec, but loading should largely work without change for WebAssembly modules. 
```

Comment:
I agree MIME type is the right thing. Is there still discussion around `.mjs` to differentiate from `.js` with bad MIME types? FWIW `.wasm`'s magic numbers can also be sniffed but ew.

Reply:
I might be totally wrong!